### PR TITLE
BUG: Fix regression on DataFrame.replace for regex

### LINKED ISSUE #25259: replace method with comma separator does not work in pandas 0.24.1
### The latest version (0.24.1) does not replace comma separator, but 0.23.4 does.

```python
import pandas as pd

df = pd.DataFrame({'one': ["1,000", "2,000"], 'two': ["3,000", "4,000"]})
df = df.replace({',': ''}, regex=True)

print(pd.__version__)
print(df)
```

```
0.23.4
    one   two
0  1000  3000
1  2000  4000
```

```
0.24.1
     one    two
0  1,000  3,000
1  2,000  4,000
```

If other characters are used with a comma, it works normally
```
df = df.replace({'1,': ''}, regex=True)
```
```
     one    two
0    000  3,000
1  2,000  4,000
```

### PR DESCRIPTION
**Pull request by PetitLepton**
The commit ensures that the replacement for regex is not confined to the 
beginning of the string but spans all the characters within. The 
behaviour is then consistent with versions prior to 0.24.0.

- [x] closes #25259
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
this seems highly related to https://github.com/pandas-dev/pandas/issues/24804 and a corresponding PR #24935 

can you have a look

**Review by jreback — CHANGES_REQUESTED** [review]
changes *always* require a test & whatnews note, but see my comment on a related issue

**Comment by charlesdong1991** [comment]
i just took a quick look... seems the problems don't look the same... shall i move whatsnew in my PR to 0.24.2 from 0.25? @jreback

**Comment by PetitLepton** [comment]
@jreback and @charlesdong1991 Hi, thanks for taking a look at this. I have very little knowledge about `pandas` internals but, from the description of #24804, it does not seem to be the same problem. I will try to include a test specifically for the case where the regex pattern is inside the string and not at the beginning. The current change already passes all the tests in `tests/frame/test_replace.py`.

**Comment by PetitLepton** [comment]
@jreback I added a test for the change and an entry in `whatsnew`. Please let me know if it matches `pandas` standard.

File: pandas/tests/frame/test_replace.py [hunk 1]
```diff
@@ -451,7 +451,11 @@ def test_regex_replace_dict_mixed(self):
 
     def test_regex_replace_dict_nested(self):
         # nested dicts will not work until this is implemented for Series
-        mix = {'a': lrange(4), 'b': list('ab..'), 'c': ['a', 'b', np.nan, 'd']}
+        mix = {
+            'a': lrange(4),
```
Comment (jreback) [inline pandas/tests/frame/test_replace.py]: can you preserve b from the original test and add another element

Comment (jreback) [reply pandas/tests/frame/test_replace.py]: add the issue number here as a comment as well

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
ideally also replicate the test from the OP as well (this could be an alternative to changing the existing test)

**Comment by WillAyd** [comment]
@PetitLepton can you address comments and merge master?

**Comment by PetitLepton** [comment]
@WillAyd @jreback My apologies for the delay. I tried to follow the recommendations, reverting the initial test to its genuine form and added a simple test referring to the issue above.

File: pandas/tests/frame/test_replace.py [hunk 2]
```diff
@@ -466,6 +466,15 @@ def test_regex_replace_dict_nested(self):
         assert_frame_equal(res3, expec)
         assert_frame_equal(res4, expec)
 
+    def test_regex_replace_dict_nested_non_first_character(self):
+        # Ensures that the regular expression is replaced
```
Comment (WillAyd) [inline pandas/tests/frame/test_replace.py]: Minor but just leave a reference to the issue number in the comments, so `# GH 25259`

File: doc/source/whatsnew/v0.24.2.rst [hunk 3]
```diff
@@ -23,6 +23,7 @@ Fixed Regressions
 - Fixed regression in :meth:`DataFrame.all` and :meth:`DataFrame.any` where ``bool_only=True`` was ignored (:issue:`25101`)
 - Fixed issue in ``DataFrame`` construction with passing a mixed list of mixed types could segfault. (:issue:`25075`)
 - Fixed regression in :meth:`DataFrame.apply` causing ``RecursionError`` when ``dict``-like classes were passed as argument. (:issue:`25196`)
+- Fixed regression in :meth:`DataFrame.replace` causing non-starting regex not being replaced. (:issue:`25259`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.2.rst]:
```suggestion
- Fixed regression in :meth:`DataFrame.replace` where ``regex=True`` was only replacing patterns matching the start of the string (:issue:`25259`)
```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor nits otherwise lgtm

File: pandas/core/internals/managers.py [hunk 4]
```diff
@@ -1901,7 +1901,7 @@ def _compare_or_regex_match(a, b, regex=False):
     """
     Compare two array_like inputs of the same shape or two scalar values
```
Comment (WillAyd) [inline pandas/core/internals/managers.py]: Since we moved from `match` to `search` here can you also updated the function name?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_replace.py (hunk 2 above)
Comment (PetitLepton) [reply pandas/tests/frame/test_replace.py]: @WillAyd Do you mean that I remove the comment and replace it only with the issue number, namely `# GH 25259`?

File: doc/source/whatsnew/v0.24.2.rst (hunk 3 above)
Comment (PetitLepton) [reply doc/source/whatsnew/v0.24.2.rst]: I am changing that now.

File: pandas/core/internals/managers.py (hunk 4 above)
Comment (PetitLepton) [reply pandas/core/internals/managers.py]: OK.

**Comment by PetitLepton** [comment]
Two tests are failing but I don't have access to the error messages.

**Comment by WillAyd** [comment]
Hmm should have access but in any case here's the error:

```sh
##[error]./pandas/core/internals/managers.py(556,48): error E128: continuation line under-indented for visual indent
```

Also if you merge master instead of rebasing shouldn't need to force push which is preferable with maintaining GitHub comment history

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @PetitLepton!

**Comment by PetitLepton** [comment]
@WillAyd and @jreback Thank you very much for your patience and guidance, it is my first (small) contribution to an open source project and I am really happy to have been able to help `pandas`. It is an amazing tool.

@TomAugspurger Thank you for approving.

**Comment by supersational** [comment]
Just spent ages confused by this behaviour, thanks for fixing it @PetitLepton !

For those who need a quick-fix the following will work around the issue:
``` python3
# add this to the start of to_replace: r"(.*?)"
# add this to the start of replacement value: r"\1"
# e.g:
df = pd.DataFrame({'first': ['abc', 'bca', 'cab']})
df.replace({r"(.*?)"+'a':  r"\1"+'.'}, regex=True)

```

**Pull request merged**